packages/core/contracts/oracle/implementation/dvm2.0/Staker.sol

```diff
@@ -0,0 +1,228 @@
+            voterStake.rewardsOutstanding;
+    }
+
+    function rewardPerToken() public view returns (uint256) {
+        if (cumulativeStaked == 0) return rewardPerTokenStored;
+
+        return rewardPerTokenStored + ((getCurrentTime() - lastUpdateTime) * emissionRate * 1e18) / cumulativeStaked;
+    }
+
+    function _updateAccountSlashingTrackers(address voterAddress) internal {
+        VoterStake storage voterStake = stakingBalances[voterAddress];
+        // Note the method below can hit a gas limit of there are a LOT of requests from the last time this was run.
+        // A future version of this should bound how many requests to look at per call to avoid gas limit issues.
+        for (uint256 i = voterStake.lastRequestIndexConsidered; i < priceRequestIds.length; i++) {
+            PriceRequest storage priceRequest = priceRequests[priceRequestIds[i].requestId];
+            VoteInstance storage voteInstance = priceRequest.voteInstances[priceRequest.lastVotingRound];
+            bytes32 revealHash = voteInstance.voteSubmissions[voterAddress].revealHash;
+
+            // The voter did not reveal or did not commit. Slash at noVote rate.
+            if (revealHash == 0)
+                voterStake.unrealizedSlash -= int256(
+                    voterStake.cumulativeStaked * requestSlashTrackers[i].noVoteSlashPerToken
+                );
+
+                // The voter did not vote with the majority. Slash at wrongVote rate.
+            else if (!voteInstance.resultComputation.wasVoteCorrect(revealHash))
+                voterStake.unrealizedSlash -= int256(
+                    voterStake.cumulativeStaked * requestSlashTrackers[i].wrongVoteSlashPerToken
+                );
+
+                // The voter voted correctly.Receive a pro-rate share of the other voters slashed amounts as a reward.
+            else {
+                uint256 roundId = rounds[priceRequestIds[i].roundId].snapshotId;
+                uint256 totalStaked = votingToken.balanceOfAt(address(this), roundId);
+                voterStake.unrealizedSlash += int256(
+                    ((voterStake.cumulativeStaked * 1e18) / totalStaked) * requestSlashTrackers[i].totalSlashed
+                );
+            }
+        }
+    }
+
+    function _updateCumulativeSlashingTrackers() internal {
+        // Note the method below can hit a gas limit of there are a LOT of requests from the last time this was run.
+        // A future version of this should bound how many requests to look at per call to avoid gas limit issues.
+        for (uint256 i = lastRequestIndexConsidered; i < priceRequestIds.length; i++) {
+            PriceRequest storage priceRequest = priceRequests[priceRequestIds[i].requestId];
+            VoteInstance storage voteInstance = priceRequest.voteInstances[priceRequest.lastVotingRound];
+            uint256 roundId = rounds[priceRequestIds[i].roundId].snapshotId;
+
+            uint256 totalStaked = votingToken.balanceOfAt(address(this), roundId);
```

Comment:
ye, I've re-worked this in a follow on commit (not yet pushed). specifically, we now need the staking contract to be snapshotable wherein we snapshot your staked amount and unrealized slashing amount on each reveal and consider this in your voting power, rather than your token balance.